src/storage-client/src/controller.rs

```diff
@@ -2714,14 +2736,25 @@ where
                         .expect("failed to connect")
                         .expect("failed to truncate write handle");
                 }
-                shard_id
+
+                // We only want to  clear shards from the finalization register
+                // if the since if actually empty.
+                if critical_since_handle.since().is_empty() {
+                    Some(shard_id)
+                } else {
+                    // TODO: periodically check shards that have not been successfully finalized.
+                    None
+                }
             })
             // Poll each future for each collection concurrently, maximum of 10 at a time.
             .buffer_unordered(10)
             // HERE BE DRAGONS: see warning on other uses of buffer_unordered
             // before any changes to `collect`
-            .collect()
-            .await;
+            .collect::<BTreeSet<Option<ShardId>>>()
```

Comment:
@guswynn You have a suggestion for bundling the `filter_map` below in this the `map` above this?